File: src/ingest.mjs
import { hashToken, nowIso } from "./db.mjs";
import { isFreshObservation, normalizeReadsbPayload, sanitizeReceiverId } from "./normalize-readsb.mjs";

function tokenHashSet(tokens) {
  return new Set((tokens || []).map((token) => hashToken(token)));
}

function getPayloadAircraft(payload) {
  if (Array.isArray(payload?.aircraft)) return payload;
  if (Array.isArray(payload?.payload?.aircraft)) return payload.payload;
  return { ...payload, aircraft: [] };
}

function optionalNumber(value) {
  if (value == null || value === "") return null;
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
}

function secondsBetween(aIso, bIso) {
  const a = Date.parse(aIso);
  const b = Date.parse(bIso);
  if (!Number.isFinite(a) || !Number.isFinite(b)) return Number.POSITIVE_INFINITY;
  return Math.abs(a - b) / 1000;
}

export function authenticateIngest(db, config, token, receiverId = null) {
  if (!token) return { ok: false, reason: "missing token" };
  const tokenHash = hashToken(token);

  const receiverToken = db
    .prepare("SELECT receiver_id FROM receiver_tokens WHERE token_hash = ?")
    .get(tokenHash);
  if (receiverToken && (!receiverId || receiverToken.receiver_id === receiverId)) {
    db.prepare("UPDATE receiver_tokens SET last_used_at = ? WHERE token_hash = ?").run(nowIso(), tokenHash);
    return { ok: true, receiverId: receiverToken.receiver_id, mode: "receiver-token" };
  }

  if (tokenHashSet(config.ingestTokens).has(tokenHash)) {
    return { ok: true, receiverId, mode: "shared-token" };
  }

  return { ok: false, reason: "invalid token" };
}

export function ingestReadsb(db, payload, options) {
  const receivedAt = options.receivedAt || nowIso();
  const receiver = payload?.receiver || {};
  const receiverId = sanitizeReceiverId(options.receiverId || receiver.id);
  if (!receiverId) {
    const error = new Error("receiver id is required");
    error.status = 400;
    throw error;
  }

  const aircraftPayload = getPayloadAircraft(payload);
  const normalized = normalizeReadsbPayload(aircraftPayload, { receivedAt });
  const receiverName = String(receiver.name || receiverId).trim().slice(0, 120);
  const publicName = String(receiver.publicName || receiver.public_name || receiverName).trim().slice(0, 120);
  const showPosition = receiver.showPosition === true || receiver.publicPosition === true;
  const receiverLat = optionalNumber(receiver.lat);
  const receiverLon = optionalNumber(receiver.lon);

  const statements = {
    upsertReceiver: db.prepare(`
      INSERT INTO receivers (
        id, name, public_name, lat, lon, show_position, last_seen_at, last_ip,
        user_agent, total_ingests, updated_at
      )
      VALUES (
        @id, @name, @publicName, @lat, @lon, @showPosition, @receivedAt, @remoteAddr,
        @userAgent, 1, @receivedAt
      )
      ON CONFLICT(id) DO UPDATE SET
        name = COALESCE(excluded.name, receivers.name),
        public_name = COALESCE(excluded.public_name, receivers.public_name),
        lat = COALESCE(excluded.lat, receivers.lat),
        lon = COALESCE(excluded.lon, receivers.lon),
        show_position = CASE WHEN excluded.show_position = 1 THEN 1 ELSE receivers.show_position END,
        last_seen_at = excluded.last_seen_at,
        last_ip = excluded.last_ip,
        user_agent = excluded.user_agent,
        total_ingests = receivers.total_ingests + 1,
        updated_at = excluded.updated_at
    `),
    insertBatch: db.prepare(`
      INSERT INTO ingest_batches (
        receiver_id, received_at, source_now, aircraft_count, accepted_count,
        track_points, remote_addr
      )
      VALUES (@receiverId, @receivedAt, @sourceNow, @aircraftCount, 0, 0, @remoteAddr)
    `),
    updateBatch: db.prepare(`
      UPDATE ingest_batches
      SET accepted_count = @acceptedCount, track_points = @trackPoints
      WHERE id = @batchId
    `),
    upsertCurrent: db.prepare(`
      INSERT INTO receiver_aircraft_current (
        receiver_id, hex, observed_at, position_at, lat, lon, flight, alt_baro,
        alt_geom, on_ground, gs, track, baro_rate, squawk, category, messages,
        rssi, seen_seconds, seen_pos_seconds, source_json, batch_id
      )
      VALUES (
        @receiverId, @hex, @observedAt, @positionAt, @lat, @lon, @flight, @altBaro,
        @altGeom, @onGround, @gs, @track, @baroRate, @squawk, @category, @messages,
        @rssi, @seenSeconds, @seenPosSeconds, @sourceJson, @batchId
      )
      ON CONFLICT(receiver_id, hex) DO UPDATE SET
        observed_at = excluded.observed_at,
        position_at = COALESCE(excluded.position_at, receiver_aircraft_current.position_at),
        lat = COALESCE(excluded.lat, receiver_aircraft_current.lat),
        lon = COALESCE(excluded.lon, receiver_aircraft_current.lon),
        flight = COALESCE(excluded.flight, receiver_aircraft_current.flight),
        alt_baro = COALESCE(excluded.alt_baro, receiver_aircraft_current.alt_baro),
        alt_geom = COALESCE(excluded.alt_geom, receiver_aircraft_current.alt_geom),
        on_ground = excluded.on_ground,
        gs = COALESCE(excluded.gs, receiver_aircraft_current.gs),
        track = COALESCE(excluded.track, receiver_aircraft_current.track),
        baro_rate = COALESCE(excluded.baro_rate, receiver_aircraft_current.baro_rate),
        squawk = COALESCE(excluded.squawk, receiver_aircraft_current.squawk),
        category = COALESCE(excluded.category, receiver_aircraft_current.category),
        messages = COALESCE(excluded.messages, receiver_aircraft_current.messages),
        rssi = COALESCE(excluded.rssi, receiver_aircraft_current.rssi),
        seen_seconds = excluded.seen_seconds,
        seen_pos_seconds = excluded.seen_pos_seconds,
        source_json = excluded.source_json,
        batch_id = excluded.batch_id
      WHERE excluded.observed_at >= receiver_aircraft_current.observed_at
    `),
    latestTrack: db.prepare(`
      SELECT position_at FROM track_points
      WHERE hex = ? AND receiver_id = ?
      ORDER BY position_at DESC
      LIMIT 1
    `),
    insertTrack: db.prepare(`
      INSERT OR IGNORE INTO track_points (
        hex, receiver_id, observed_at, position_at, lat, lon, alt_baro, alt_geom,
        on_ground, gs, track, messages, batch_id
      )
      VALUES (
        @hex, @receiverId, @observedAt, @positionAt, @lat, @lon, @altBaro, @altGeom,
        @onGround, @gs, @track, @messages, @batchId
      )
    `),
    upsertSighting: db.prepare(`
      INSERT INTO aircraft_sightings (hex, first_seen_at, last_seen_at, total_observations, last_flight)
      VALUES (@hex, @observedAt, @observedAt, 1, @flight)
      ON CONFLICT(hex) DO UPDATE SET
        last_seen_at = CASE
          WHEN excluded.last_seen_at > aircraft_sightings.last_seen_at
          THEN excluded.last_seen_at
          ELSE aircraft_sightings.last_seen_at
        END,
        total_observations = aircraft_sightings.total_observations + 1,
        last_flight = COALESCE(excluded.last_flight, aircraft_sightings.last_flight)
    `),
  };

  const tx = db.transaction(() => {
    statements.upsertReceiver.run({
      id: receiverId,
      name: receiverName || receiverId,
      publicName: publicName || receiverName || receiverId,
      lat: receiverLat,
      lon: receiverLon,
      showPosition: showPosition ? 1 : 0,
      receivedAt,
      remoteAddr: options.remoteAddr || null,
      userAgent: options.userAgent || null,
    });

    const batch = statements.insertBatch.run({
      receiverId,
      receivedAt,
      sourceNow: normalized.sourceNow,
      aircraftCount: aircraftPayload.aircraft.length,
      remoteAddr: options.remoteAddr || null,
    });

    let acceptedCount = 0;
    let trackPoints = 0;
    let staleCount = 0;
    const changedHexes = new Set();

    for (const observation of normalized.aircraft) {
      if (!isFreshObservation(observation, receivedAt, options.maxObservationAgeSeconds)) {
        staleCount += 1;
        continue;
      }

      const row = {
        receiverId,
        batchId: batch.lastInsertRowid,
        hex: observation.hex,
        observedAt: observation.observedAt,
        positionAt: observation.positionAt,
        lat: observation.lat,
        lon: observation.lon,
        flight: observation.flight,
        altBaro: observation.altBaro,
        altGeom: observation.altGeom,
        onGround: observation.onGround ? 1 : 0,
        gs: observation.gs,
        track: observation.track,
        baroRate: observation.baroRate,
        squawk: observation.squawk,
        category: observation.category,
        messages: observation.messages,
        rssi: observation.rssi,
        seenSeconds: observation.seenSeconds,
        seenPosSeconds: observation.seenPosSeconds,
        sourceJson: JSON.stringify(observation.raw),
      };

      statements.upsertCurrent.run(row);
      statements.upsertSighting.run(row);
      acceptedCount += 1;
      changedHexes.add(observation.hex);

      if (row.lat != null && row.lon != null && row.positionAt) {
        const latest = statements.latestTrack.get(row.hex, receiverId);
        if (!latest || secondsBetween(latest.position_at, row.positionAt) >= options.trackMinIntervalSeconds) {
          trackPoints += statements.insertTrack.run(row).changes;
        }
      }
    }

    statements.updateBatch.run({
      batchId: batch.lastInsertRowid,
      acceptedCount,
      trackPoints,
    });

    return {
      receiverId,
      batchId: batch.lastInsertRowid,
      receivedAt,
      sourceNow: normalized.sourceNow,
      aircraftCount: aircraftPayload.aircraft.length,
      acceptedCount,
      staleCount,
      trackPoints,
      changedHexes: [...changedHexes],
    };
  });

  return tx();
}

export function getCurrentAircraft(db, options = {}) {
  const now = options.now || nowIso();
  const cutoff = new Date(Date.parse(now) - options.currentWindowSeconds * 1000).toISOString();
  const rows = db
    .prepare(`
      SELECT
        c.*,
        r.public_name AS receiver_public_name
      FROM receiver_aircraft_current c
      JOIN receivers r ON r.id = c.receiver_id
      WHERE c.observed_at >= ?
      ORDER BY c.hex, c.observed_at DESC
    `)
    .all(cutoff);

  const groups = new Map();
  for (const row of rows) {
    const group = groups.get(row.hex) || [];
    group.push(row);
    groups.set(row.hex, group);
  }

  const aircraft = [];
  for (const [hex, group] of groups) {
    group.sort((a, b) => {
      const aPos = a.lat != null && a.lon != null ? 1 : 0;
      const bPos = b.lat != null && b.lon != null ? 1 : 0;
      if (aPos !== bPos) return bPos - aPos;
      return String(b.position_at || b.observed_at).localeCompare(String(a.position_at || a.observed_at));
    });
    const best = group[0];
    const receiverIds = new Set(group.map((row) => row.receiver_id));
    aircraft.push({
      hex,
      flight: best.flight,
      lat: best.lat,
      lon: best.lon,
      altBaro: best.alt_baro,
      altGeom: best.alt_geom,
      onGround: Boolean(best.on_ground),
      gs: best.gs,
      track: best.track,
      baroRate: best.baro_rate,
      squawk: best.squawk,
      category: best.category,
      messages: best.messages,
      rssi: best.rssi,
      observedAt: best.observed_at,
      positionAt: best.position_at,
      receiverCount: receiverIds.size,
      bestReceiverId: best.receiver_id,
      receivers: [...receiverIds].sort(),
    });
  }

  aircraft.sort((a, b) => {
    const aPos = a.lat != null && a.lon != null ? 1 : 0;
    const bPos = b.lat != null && b.lon != null ? 1 : 0;
    if (aPos !== bPos) return bPos - aPos;
    return String(a.flight || a.hex).localeCompare(String(b.flight || b.hex));
  });

  return {
    now,
    cutoff,
    count: aircraft.length,
    aircraft,
  };
}

export function getPublicReceivers(db, options = {}) {
  const now = options.now || nowIso();
  const cutoff = new Date(Date.parse(now) - options.currentWindowSeconds * 1000).toISOString();
  const rows = db
    .prepare(`
      SELECT
        r.id,
        r.public_name,
        r.lat,
        r.lon,
        r.show_position,
        r.last_seen_at,
        r.total_ingests,
        COUNT(c.hex) AS current_aircraft
      FROM receivers r
      LEFT JOIN receiver_aircraft_current c
        ON c.receiver_id = r.id AND c.observed_at >= ?
      GROUP BY r.id
      ORDER BY r.public_name, r.id
    `)
    .all(cutoff);

  return rows.map((row) => ({
    id: row.id,
    name: row.public_name || row.id,
    online: row.last_seen_at ? Date.parse(now) - Date.parse(row.last_seen_at) <= options.currentWindowSeconds * 1000 : false,
    lastSeenAt: row.last_seen_at,
    totalIngests: row.total_ingests,
    currentAircraft: row.current_aircraft,
    lat: row.show_position ? row.lat : null,
    lon: row.show_position ? row.lon : null,
  }));
}

export function getTrack(db, hex, options = {}) {
  const normalizedHex = String(hex || "").trim().toLowerCase();
  if (!/^[0-9a-f]{6}$/.test(normalizedHex)) return [];
  const nowMs = Date.parse(options.now || nowIso());
  const to = options.to ? new Date(options.to).toISOString() : new Date(nowMs).toISOString();
  const from = options.from
    ? new Date(options.from).toISOString()
    : new Date(nowMs - 24 * 60 * 60 * 1000).toISOString();
  const limit = Math.max(1, Math.min(options.limit || 10000, 50000));

  return db
    .prepare(`
      SELECT
        hex, receiver_id AS receiverId, observed_at AS observedAt,
        position_at AS positionAt, lat, lon, alt_baro AS altBaro,
        alt_geom AS altGeom, on_ground AS onGround, gs, track, messages
      FROM track_points
      WHERE hex = ? AND position_at >= ? AND position_at <= ?
      ORDER BY position_at ASC
      LIMIT ?
    `)
    .all(normalizedHex, from, to, limit)
    .map((row) => ({ ...row, onGround: Boolean(row.onGround) }));
}

File: server/ingest/ingest.go
package ingest

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"io"
	"math"
	"sort"
	"strings"
	"time"

	"github.com/luftaquila/skytrace/server/database"
)

const (
	maxIngestAircraft          = 1000
	maxReceiverCurrentAircraft = 20000
)

type AuthResult struct {
	OK         bool
	Reason     string
	ReceiverID string
	TokenHash  string
}

type Options struct {
	ReceiverID               string
	TokenHash                string
	ReceivedAt               time.Time
	RemoteAddr               string
	UserAgent                string
	MaxObservationAgeSeconds float64
	TrackMinIntervalSeconds  float64
	PositionFilterMaxMach    float64
	ConsumeTrackBudget       func(string) bool
}

type Result struct {
	ReceiverID                  string   `json:"receiverId"`
	BatchID                     int64    `json:"batchId"`
	ReceivedAt                  string   `json:"receivedAt"`
	SourceNow                   *string  `json:"sourceNow"`
	AircraftCount               int      `json:"aircraftCount"`
	TruncatedCount              int      `json:"truncatedCount"`
	AcceptedCount               int      `json:"acceptedCount"`
	InvalidObservationCount     int      `json:"invalidObservationCount"`
	InvalidFieldCount           int      `json:"invalidFieldCount"`
	TruncatedFieldCount         int      `json:"truncatedFieldCount"`
	FilteredPositionCount       int      `json:"filteredPositionCount"`
	CurrentCapacityDroppedCount int      `json:"currentCapacityDroppedCount"`
	TrackBudgetDroppedCount     int      `json:"trackBudgetDroppedCount"`
	TrackPoints                 int      `json:"trackPoints"`
	ChangedHexes                []string `json:"changedHexes"`
}

func DecodePayload(bytes []byte) (map[string]any, error) {
	decoder := json.NewDecoder(strings.NewReader(string(bytes)))
	decoder.UseNumber()
	var payload map[string]any
	if err := decoder.Decode(&payload); err != nil {
		return nil, err
	}
	if payload == nil {
		payload = map[string]any{}
	}
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return nil, errors.New("multiple JSON values")
	}
	return payload, nil
}

func Authenticate(ctx context.Context, db *sql.DB, token, receiverID string) (AuthResult, error) {
	if token == "" {
		return AuthResult{Reason: "missing token"}, nil
	}
	hash := database.HashToken(token)
	var configuredReceiver string
	err := db.QueryRowContext(
		ctx,
		"SELECT receiver_id FROM receiver_tokens WHERE token_hash = ?",
		hash,
	).Scan(&configuredReceiver)
	if errors.Is(err, sql.ErrNoRows) {
		return AuthResult{Reason: "invalid token"}, nil
	}
	if err != nil {
		return AuthResult{}, err
	}
	if receiverID != "" && configuredReceiver != receiverID {
		return AuthResult{Reason: "invalid token"}, nil
	}
	return AuthResult{
		OK:         true,
		ReceiverID: configuredReceiver,
		TokenHash:  hash,
	}, nil
}

func Store(ctx context.Context, db *sql.DB, payload map[string]any, options Options) (Result, error) {
	receiverID, ok := SanitizeReceiverID(options.ReceiverID)
	if !ok {
		return Result{}, errors.New("receiver id is required")
	}
	if options.ReceivedAt.IsZero() {
		options.ReceivedAt = time.Now()
	}
	if !isFinite(options.MaxObservationAgeSeconds) {
		options.MaxObservationAgeSeconds = 120
	}
	if !isFinite(options.PositionFilterMaxMach) || options.PositionFilterMaxMach == 0 {
		options.PositionFilterMaxMach = 3.5
	}
	receivedAt := iso(options.ReceivedAt)
	aircraftPayload, rawAircraft := selectedAircraftPayload(payload)
	boundedAircraft := rawAircraft
	if len(boundedAircraft) > maxIngestAircraft {
		boundedAircraft = boundedAircraft[:maxIngestAircraft]
	}
	aircraftPayload["aircraft"] = boundedAircraft
	normalized := NormalizePayload(aircraftPayload, options.ReceivedAt, options.MaxObservationAgeSeconds)

	receiver, _ := payload["receiver"].(map[string]any)
	receiverName, nameInvalid, nameTruncated := boundedText(receiver["name"], 120, receiverID)
	publicRaw, exists := receiver["publicName"]
	if !exists {
		publicRaw = receiver["public_name"]
	}
	publicName, publicInvalid, publicTruncated := boundedText(publicRaw, 120, receiverName)
	remoteAddr, remoteInvalid, remoteTruncated := boundedText(options.RemoteAddr, 64, "")
	userAgent, agentInvalid, agentTruncated := boundedText(options.UserAgent, 256, "")
	receiverLat, latValid := coordinate(receiver["lat"], -90, 90)
	receiverLon, lonValid := coordinate(receiver["lon"], -180, 180)
	invalidFields := normalized.InvalidFieldCount
	if receiver != nil {
		if value, exists := receiver["lat"]; exists && value != nil && !latValid {
			invalidFields++
		}
		if value, exists := receiver["lon"]; exists && value != nil && !lonValid {
			invalidFields++
		}
	}
	for _, invalid := range []bool{nameInvalid, publicInvalid, remoteInvalid, agentInvalid} {
		if invalid {
			invalidFields++
		}
	}
	truncatedFields := normalized.TruncatedFieldCount
	for _, truncated := range []bool{nameTruncated, publicTruncated, remoteTruncated, agentTruncated} {
		if truncated {
			truncatedFields++
		}
	}

	transaction, release, err := database.WriteTx(ctx, db)
	if err != nil {
		return Result{}, err
	}
	defer release()
	defer transaction.Rollback()

	if _, err := transaction.ExecContext(ctx, `
		INSERT INTO receivers (
		  id, name, public_name, lat, lon, last_seen_at, last_ip,
		  user_agent, total_ingests, updated_at
		)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, 1, ?)
		ON CONFLICT(id) DO UPDATE SET
		  name = COALESCE(excluded.name, receivers.name),
		  public_name = COALESCE(excluded.public_name, receivers.public_name),
		  lat = COALESCE(excluded.lat, receivers.lat),
		  lon = COALESCE(excluded.lon, receivers.lon),
		  last_seen_at = excluded.last_seen_at,
		  last_ip = excluded.last_ip,
		  user_agent = excluded.user_agent,
		  total_ingests = receivers.total_ingests + 1,
		  updated_at = excluded.updated_at
	`, receiverID, receiverName, publicName, receiverLat, receiverLon, receivedAt,
		nullableText(remoteAddr), nullableText(userAgent), receivedAt); err != nil {
		return Result{}, err
	}

	batch, err := transaction.ExecContext(ctx, `
		INSERT INTO ingest_batches (
		  receiver_id, received_at, source_now, aircraft_count, accepted_count,
		  track_points, remote_addr
		) VALUES (?, ?, ?, ?, 0, 0, ?)
	`, receiverID, receivedAt, normalized.SourceNow, len(rawAircraft), nullableText(remoteAddr))
	if err != nil {
		return Result{}, err
	}
	batchID, err := batch.LastInsertId()
	if err != nil {
		return Result{}, err
	}

	var currentCount int
	if err := transaction.QueryRowContext(
		ctx,
		"SELECT count(*) FROM receiver_aircraft_current WHERE receiver_id = ?",
		receiverID,
	).Scan(&currentCount); err != nil {
		return Result{}, err
	}
	result := Result{
		ReceiverID:              receiverID,
		BatchID:                 batchID,
		ReceivedAt:              receivedAt,
		SourceNow:               normalized.SourceNow,
		AircraftCount:           len(rawAircraft),
		TruncatedCount:          max(0, len(rawAircraft)-maxIngestAircraft),
		InvalidObservationCount: normalized.InvalidObservationCount,
		InvalidFieldCount:       invalidFields,
		TruncatedFieldCount:     truncatedFields,
		ChangedHexes:            []string{},
	}
	changed := make(map[string]struct{})
	for _, observation := range normalized.Aircraft {
		var exists int
		err := transaction.QueryRowContext(
			ctx,
			"SELECT 1 FROM receiver_aircraft_current WHERE receiver_id = ? AND hex = ?",
			receiverID,
			observation.Hex,
		).Scan(&exists)
		if err != nil && !errors.Is(err, sql.ErrNoRows) {
			return Result{}, err
		}
		present := err == nil
		if !present && currentCount >= maxReceiverCurrentAircraft {
			result.CurrentCapacityDroppedCount++
			continue
		}

		var latest *trackPosition
		if observation.Lat != nil && observation.Lon != nil && observation.PositionAt != nil {
			var value trackPosition
			err := transaction.QueryRowContext(ctx, `
				SELECT position_at, lat, lon FROM track_points
				WHERE hex = ? AND receiver_id = ?
				ORDER BY id DESC
				LIMIT 1
			`, observation.Hex, receiverID).Scan(&value.PositionAt, &value.Lat, &value.Lon)
			if err != nil && !errors.Is(err, sql.ErrNoRows) {
				return Result{}, err
			}
			if err == nil {
				latest = &value
			}
		}
		if latest != nil && !plausiblePosition(*latest, observation, options.PositionFilterMaxMach) {
			observation.PositionAt = nil
			observation.Lat = nil
			observation.Lon = nil
			result.FilteredPositionCount++
		}

		if err := upsertCurrent(ctx, transaction, receiverID, observation); err != nil {
			return Result{}, err
		}
		if !present {
			currentCount++
		}
		result.AcceptedCount++
		if _, exists := changed[observation.Hex]; !exists {
			changed[observation.Hex] = struct{}{}
			result.ChangedHexes = append(result.ChangedHexes, observation.Hex)
		}

		if observation.Lat != nil && observation.Lon != nil && observation.PositionAt != nil {
			shouldTrack := latest == nil || secondsBetween(latest.PositionAt, *observation.PositionAt) >= options.TrackMinIntervalSeconds
			if shouldTrack {
				allowed := options.ConsumeTrackBudget == nil || options.ConsumeTrackBudget(receiverID)
				if !allowed {
					result.TrackBudgetDroppedCount++
				} else {
					changes, err := insertTrack(ctx, transaction, receiverID, observation)
					if err != nil {
						return Result{}, err
					}
					result.TrackPoints += changes
				}
			}
		}
	}
	if _, err := transaction.ExecContext(
		ctx,
		"UPDATE ingest_batches SET accepted_count = ?, track_points = ? WHERE id = ?",
		result.AcceptedCount,
		result.TrackPoints,
		batchID,
	); err != nil {
		return Result{}, err
	}
	if options.TokenHash != "" {
		if _, err := transaction.ExecContext(
			ctx,
			"UPDATE receiver_tokens SET last_used_at = ? WHERE token_hash = ?",
			receivedAt,
			options.TokenHash,
		); err != nil {
			return Result{}, err
		}
	}
	if err := transaction.Commit(); err != nil {
		return Result{}, err
	}
	return result, nil
}

func selectedAircraftPayload(payload map[string]any) (map[string]any, []any) {
	if aircraft, ok := payload["aircraft"].([]any); ok {
		clone := cloneMap(payload)
		return clone, aircraft
	}
	if nested, ok := payload["payload"].(map[string]any); ok {
		if aircraft, ok := nested["aircraft"].([]any); ok {
			clone := cloneMap(nested)
			return clone, aircraft
		}
	}
	clone := cloneMap(payload)
	return clone, nil
}

func cloneMap(source map[string]any) map[string]any {
	clone := make(map[string]any, len(source)+1)
	for key, value := range source {
		clone[key] = value
	}
	return clone
}

func boundedText(value any, limit int, fallback string) (string, bool, bool) {
	if value == nil {
		return fallback, false, false
	}
	text, ok := scalarString(value)
	if !ok {
		return fallback, true, false
	}
	text = strings.TrimSpace(text)
	if text == "" {
		return fallback, false, false
	}
	runes := []rune(text)
	truncated := len(runes) > limit
	if truncated {
		text = string(runes[:limit])
	}
	return text, false, truncated
}

func coordinate(value any, minValue, maxValue float64) (*float64, bool) {
	if value == nil || value == "" {
		return nil, true
	}
	number, ok := jsNumber(value)
	if !ok || !isFinite(number) || number < minValue || number > maxValue {
		return nil, false
	}
	return &number, true
}

type trackPosition struct {
	PositionAt string
	Lat        float64
	Lon        float64
}

func plausiblePosition(previous trackPosition, current Observation, maxMach float64) bool {
	if current.Lat == nil || current.Lon == nil || current.PositionAt == nil {
		return true
	}
	hours := secondsBetween(previous.PositionAt, *current.PositionAt) / 3600
	if !isFinite(hours) || hours <= 0 {
		return false
	}
	distance := distanceNauticalMiles(previous.Lat, previous.Lon, *current.Lat, *current.Lon)
	return distance/hours <= maxMach*666.739
}

func secondsBetween(first, second string) float64 {
	a, err := time.Parse(time.RFC3339Nano, first)
	if err != nil {
		return math.Inf(1)
	}
	b, err := time.Parse(time.RFC3339Nano, second)
	if err != nil {
		return math.Inf(1)
	}
	return b.Sub(a).Seconds()
}

func distanceNauticalMiles(aLat, aLon, bLat, bLon float64) float64 {
	const radiusNM = 3440.065
	toRadians := func(value float64) float64 { return value * math.Pi / 180 }
	dLat := toRadians(bLat - aLat)
	dLon := toRadians(bLon - aLon)
	lat1 := toRadians(aLat)
	lat2 := toRadians(bLat)
	h := math.Pow(math.Sin(dLat/2), 2) +
		math.Cos(lat1)*math.Cos(lat2)*math.Pow(math.Sin(dLon/2), 2)
	return 2 * radiusNM * math.Asin(math.Min(1, math.Sqrt(h)))
}

func upsertCurrent(ctx context.Context, transaction *sql.Tx, receiverID string, value Observation) error {
	_, err := transaction.ExecContext(ctx, `
		INSERT INTO receiver_aircraft_current (
		  receiver_id, hex, observed_at, position_at, lat, lon, flight, alt_baro,
		  alt_geom, on_ground, gs, ias, tas, mach, track, true_heading, mag_heading,
		  baro_rate, geom_rate, track_rate, roll, squawk, category, source_type,
		  source_kind, emergency, nav_qnh, nav_altitude_mcp, nav_altitude_fms,
		  nav_heading, wd, ws, oat, tat, nac_p, nac_v, nic, nic_baro, rc, sil,
		  sil_type, version, alert, spi, non_icao, messages, rssi, seen_seconds,
		  seen_pos_seconds
		) VALUES (
		  ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,
		  ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?
		)
		ON CONFLICT(receiver_id, hex) DO UPDATE SET
		  observed_at=excluded.observed_at, position_at=excluded.position_at,
		  lat=excluded.lat, lon=excluded.lon, flight=excluded.flight,
		  alt_baro=excluded.alt_baro, alt_geom=excluded.alt_geom, on_ground=excluded.on_ground,
		  gs=excluded.gs, ias=excluded.ias, tas=excluded.tas, mach=excluded.mach,
		  track=excluded.track, true_heading=excluded.true_heading, mag_heading=excluded.mag_heading,
		  baro_rate=excluded.baro_rate, geom_rate=excluded.geom_rate,
		  track_rate=excluded.track_rate, roll=excluded.roll, squawk=excluded.squawk,
		  category=excluded.category, source_type=excluded.source_type,
		  source_kind=excluded.source_kind, emergency=excluded.emergency,
		  nav_qnh=excluded.nav_qnh, nav_altitude_mcp=excluded.nav_altitude_mcp,
		  nav_altitude_fms=excluded.nav_altitude_fms, nav_heading=excluded.nav_heading,
		  wd=excluded.wd, ws=excluded.ws, oat=excluded.oat, tat=excluded.tat,
		  nac_p=excluded.nac_p, nac_v=excluded.nac_v, nic=excluded.nic,
		  nic_baro=excluded.nic_baro, rc=excluded.rc, sil=excluded.sil,
		  sil_type=excluded.sil_type, version=excluded.version, alert=excluded.alert,
		  spi=excluded.spi, non_icao=excluded.non_icao, messages=excluded.messages,
		  rssi=excluded.rssi, seen_seconds=excluded.seen_seconds,
		  seen_pos_seconds=excluded.seen_pos_seconds
		WHERE excluded.observed_at >= receiver_aircraft_current.observed_at
	`,
		receiverID, value.Hex, value.ObservedAt, value.PositionAt, value.Lat, value.Lon,
		value.Flight, value.AltBaro, value.AltGeom, boolInteger(value.OnGround), value.GS,
		value.IAS, value.TAS, value.Mach, value.Track, value.TrueHeading, value.MagHeading,
		value.BaroRate, value.GeomRate, value.TrackRate, value.Roll, value.Squawk,
		value.Category, value.SourceType, value.SourceKind, value.Emergency, value.NavQNH,
		value.NavAltitudeMCP, value.NavAltitudeFMS, value.NavHeading, value.WindDirection,
		value.WindSpeed, value.OAT, value.TAT, value.NACP, value.NACV, value.NIC,
		value.NICBaro, value.RC, value.SIL, value.SILType, value.Version, value.Alert,
		value.SPI, boolInteger(value.NonICAO), value.Messages, value.RSSI, value.SeenSeconds,
		value.SeenPosSeconds,
	)
	return err
}

func insertTrack(ctx context.Context, transaction *sql.Tx, receiverID string, value Observation) (int, error) {
	result, err := transaction.ExecContext(ctx, `
		INSERT OR IGNORE INTO track_points (
		  hex, receiver_id, observed_at, position_at, lat, lon, alt_baro, alt_geom,
		  on_ground, gs, ias, tas, mach, track, true_heading, mag_heading,
		  baro_rate, geom_rate, wd, ws, oat, tat, source_type, messages, rssi
		) VALUES (
		  ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?
		)
	`, value.Hex, receiverID, value.ObservedAt, value.PositionAt, value.Lat, value.Lon,
		value.AltBaro, value.AltGeom, boolInteger(value.OnGround), value.GS, value.IAS,
		value.TAS, value.Mach, value.Track, value.TrueHeading, value.MagHeading,
		value.BaroRate, value.GeomRate, value.WindDirection, value.WindSpeed, value.OAT,
		value.TAT, value.SourceType, value.Messages, value.RSSI)
	if err != nil {
		return 0, err
	}
	changes, err := result.RowsAffected()
	return int(changes), err
}

func boolInteger(value bool) int {
	if value {
		return 1
	}
	return 0
}

func nullableText(value string) any {
	if value == "" {
		return nil
	}
	return value
}

func SortedChangedHexes(values []string) []string {
	copyOfValues := append([]string(nil), values...)
	sort.Strings(copyOfValues)
	return copyOfValues
}
